pkg/volumes/alicloud/discovery.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alicloud
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/golang/glog"
+	"kope.io/etcd-manager/pkg/privateapi/discovery"
+	"kope.io/etcd-manager/pkg/volumes"
+)
+
+// AlicloudVolumes allows us to discover our peer nodes
+var _ discovery.Interface = &AlicloudVolumes{}
+
+func (a *AlicloudVolumes) Poll() (map[string]discovery.Node, error) {
+	nodes := make(map[string]discovery.Node)
+
+	allVolumes, err := a.findVolumes(false)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceToVolumeMap := make(map[string]*volumes.Volume)
+	for _, v := range allVolumes {
+		if v.AttachedTo != "" {
+			instanceToVolumeMap[v.AttachedTo] = v
+		}
+	}
+
+	if len(instanceToVolumeMap) != 0 {
+		request := ecs.CreateDescribeInstancesRequest()
+		request.RegionId = a.region
+		request.PageSize = requests.NewInteger(PageSizeLarge)
+
+		var ids []string
+		for id := range instanceToVolumeMap {
+			ids = append(ids, id)
+		}
+		request.InstanceIds = "[\"" + strings.Join(ids, "\",\"") + "\"]"
+
+		response, err := a.ecs.DescribeInstances(request)
+		if err != nil {
+			return nil, fmt.Errorf("error from Alicloud DescribeInstances: %v", err)
+		}
+
+		for _, instance := range response.Instances.Instance {
+			volume := instanceToVolumeMap[instance.InstanceId]
+			if volume == nil {
+				// unexpected ... we constructed the request from the map!
+				glog.Errorf("instance not found: %q", instance.InstanceId)
+				continue
+			}
+
+			// We use the etcd node ID as the persistent identifier, because the data determines who we are
+			node := discovery.Node{
+				ID: volume.EtcdName,
+			}
+			privateIPs := instance.InnerIpAddress.IpAddress
```

Comment:
what is difference between inneripaddress and networkinterfaces?

Reply:
As described in the API docs, NetworkInterfaces contain Elastic Network Interfaces, while InnerIpAddress contains the Private IP address of the instance. 

I know you might not like the term 'InnerIpAddress', but it is safe to use this one for Private IP address. 😉